crypto/bls_test.go

```diff
@@ -105,6 +106,17 @@ func TestBLSEncodeDecode(t *testing.T) {
 	require.Error(t, err, "the key decoding should fail - key value is identity")
 	assert.IsType(t, expectedError, err)
 }
+
+// TestBLSInvalidSignature tests the function BLSInvalidSignature
+func TestBLSInvalidSignature(t *testing.T) {
```

Comment:
If we call `Verify(anyMessage, BLSInvalidSignature(), anyKey)`, it will return `false, nil` and not `false, someError` right? Could we add a test for this?

Reply:
I'm afraid it won't (i.e. the verification will return an Error). You'd return `RLC_ERR` [here](https://github.com/onflow/flow-go/blob/6c4685b73e99206b199bd3c5ccc7dd7ced7c8517/crypto/bls_core.c#L319-L322) and fail [here](https://github.com/onflow/flow-go/blob/6c4685b73e99206b199bd3c5ccc7dd7ced7c8517/crypto/bls.go#L156-L159) on the go side.

But considering signatures are externally-controlled, that's always an eventuality, and errors in verification should hence *always* be handled. Is there a situation where we do not?

Reply:
Known errors are handled by error type. An unknown error type will crash HotStuff ([this](https://github.com/onflow/flow-go/blob/a603c086d66b5e51e73c39d6ab6485f70d710cd8/consensus/hotstuff/validator/validator.go#L146-L157) is the relevant error handling code for validating votes). 

⚠️  From [your link here](https://github.com/onflow/flow-go/blob/6c4685b73e99206b199bd3c5ccc7dd7ced7c8517/crypto/bls.go#L156-L159), which simply returns `fmt.Errorf(...)` it _looks_ like this case will not return a well-defined error type, and consequently would not be handled.

Reply:
Sorry I'm late to this!
The current implementation would return `false, nil`. Any invalid signature wouldn't error, the verification function just notifies that the signature can't authenticate the public key owner, but it doesn't provide more info. 

The verification errors in 2 cases:
 - a sentinel error for invalid inputs, when the developer enters invalid inputs that are not related to the signature (nil hasher for instance) 
 - a random error if something unexpected happens (a memory issue for instance).
 
 I'll add tests to the PR to make this clearer.

Reply:
🙀 If somebody comes after me: [those](https://github.com/onflow/flow-go/blob/6c4685b73e99206b199bd3c5ccc7dd7ced7c8517/crypto/bls12381_utils.h#L14-L15) are the important lines I missed.
Sorry for the wrong info, @jordanschalm. Thanks for the input on the downstream handling, which is another kettle of fish, I guess. 
Thanks for the correction and the tests, @tarakby